### PR TITLE
make kube-aggregator run as static pod for local-up-cluster

### PR DESCRIPTION
**Pull request by deads2k**
Runs the kube-aggregator as a static pod for local-up-cluster.  Looks like someone broke kubectl negotiation again, so I'll fix that up separately.

@kubernetes/sig-api-machinery-misc 
@lavalamp you're probably looking to run kube-aggregator as a static pod, here's an example.
@jwforres I'll make a secure variant for wiring up to openshift.

### CODE REVIEW

**Comment by deads2k** [comment]
@sttts bump

**Comment by deads2k** [comment]
@foxish ptal

File: cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    - "--core-kubeconfig=/var/run/auth-client/kube-aggregator.kubeconfig"
+    - "--authentication-kubeconfig=/var/run/auth-client/kube-aggregator.kubeconfig"
+    - "--authorization-kubeconfig=/var/run/auth-client/kube-aggregator.kubeconfig"
+    - "--proxy-client-cert-file=/var/run/auth-proxy-client/client-auth-proxy.crt"
+    - "--proxy-client-key-file=/var/run/auth-proxy-client/client-auth-proxy.key"
+    - "--tls-cert-file=/var/run/serving-cert/serving-kube-aggregator.crt"
+    - "--tls-private-key-file=/var/run/serving-cert/serving-kube-aggregator.key"
+    - "--tls-ca-file=/var/run/serving-ca/server-ca.crt"
+    - "--client-ca-file=/var/run/client-ca/client-ca.crt"
+    - "--requestheader-username-headers=X-Remote-User"
```
Comment (sttts) [inline cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml]: aren't there sensible defaults for these?

File: cmd/kube-aggregator/pkg/cmd/server/start.go [hunk 2]
```diff
@@ -93,6 +98,9 @@ func NewCommandStartDiscoveryServer(out, err io.Writer) *cobra.Command {
 	o.Authorization.AddFlags(flags)
 	flags.StringVar(&o.ProxyClientCertFile, "proxy-client-cert-file", o.ProxyClientCertFile, "client certificate used identify the proxy to the API server")
 	flags.StringVar(&o.ProxyClientKeyFile, "proxy-client-key-file", o.ProxyClientKeyFile, "client certificate key used identify the proxy to the API server")
+	flags.StringVar(&o.CoreAPIKubeconfig, "core-kubeconfig", o.CoreAPIKubeconfig, ""+
+		"kubeconfig file pointing at the 'core' kubernetes server with enough rights to get,list,watch "+
+		" services,endpoints.  If not set, the in cluster config is used")
```
Comment (sttts) [inline cmd/kube-aggregator/pkg/cmd/server/start.go]: in-cluster

File: hack/local-up-cluster.sh [hunk 3]
```diff
@@ -476,6 +495,10 @@ function start_controller_manager {
 
 function start_kubelet {
     KUBELET_LOG=/tmp/kubelet.log
+    STATIC_POD_DIR=/tmp/local-up-static-pods
```
Comment (sttts) [inline hack/local-up-cluster.sh]: Why not in /var/run/kubernetes?

File: hack/local-up-cluster.sh [hunk 4]
```diff
@@ -392,9 +398,13 @@ function start_apiserver {
     kube::util::create_client_certkey "${CONTROLPLANE_SUDO}" "${CERT_DIR}" 'client-ca' scheduler system:scheduler system:masters
     kube::util::create_client_certkey "${CONTROLPLANE_SUDO}" "${CERT_DIR}" 'client-ca' admin system:admin system:masters
 
-    # Create auth proxy client ca
-    kube::util::create_signing_certkey "${CONTROLPLANE_SUDO}" "${CERT_DIR}" request-header '"client auth"'
+    # Create matching certificates for kube-aggregator
+    kube::util::create_serving_certkey "${CONTROLPLANE_SUDO}" "${CERT_DIR}" "server-ca" kube-aggregator api.kube-public.svc "localhost" ${API_HOST_IP}
```
Comment (sttts) [inline hack/local-up-cluster.sh]: This is getting complicated.

Comment (deads2k) [inline hack/local-up-cluster.sh]: > This is getting complicated.

Looks about the same as signing for the apiserver.

File: cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml (hunk 1 above)
Comment (deads2k) [inline cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml]: > aren't there sensible defaults for these?

Maybe?  Reassess at another time?

File: hack/local-up-cluster.sh (hunk 3 above)
Comment (deads2k) [inline hack/local-up-cluster.sh]: > Why not in /var/run/kubernetes?

I don't feel strongly.  `/var/run/kubernetes/static-pods`?

File: cmd/kube-aggregator/pkg/cmd/server/start.go [hunk 5]
```diff
@@ -138,10 +146,18 @@ func (o DiscoveryServerOptions) RunDiscoveryServer() error {
...
-		return err
+	var kubeconfig *rest.Config
+	if len(o.CoreAPIKubeconfig) > 0 {
+		loadingRules := &clientcmd.ClientConfigLoadingRules{ExplicitPath: o.CoreAPIKubeconfig}
+		loader := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &clientcmd.ConfigOverrides{})
+
+		kubeconfig, err = loader.ClientConfig()
+
+	} else {
+		// without the remote kubeconfig file, try to use the in-cluster config.  Most addon API servers will use this path
```
Comment (sttts) [inline cmd/kube-aggregator/pkg/cmd/server/start.go]: no need for the comment

File: hack/local-up-cluster.sh [hunk 6]
```diff
@@ -450,6 +462,13 @@ function start_apiserver {
             AUTH_ARGS="--username=admin --password=admin"
         fi
     fi
+
+    # create the kube-public namespace for the aggregator
+    ${KUBECTL} --kubeconfig="${CERT_DIR}/admin.kubeconfig" create namespace kube-public
```
Comment (sttts) [inline hack/local-up-cluster.sh]: why is this needed?

File: hack/local-up-cluster.sh [hunk 7]
```diff
@@ -450,6 +462,13 @@ function start_apiserver {
...
+    ${CONTROLPLANE_SUDO} cp "${CERT_DIR}/admin.kubeconfig" "${CERT_DIR}/admin-kube-aggregator.kubeconfig"
+    ${CONTROLPLANE_SUDO} chown $(whoami) "${CERT_DIR}/admin-kube-aggregator.kubeconfig"
+    ${KUBECTL} config set-cluster local-up-cluster --kubeconfig="${CERT_DIR}/admin-kube-aggregator.kubeconfig" --server="https://${API_HOST_IP}:9443"
```
Comment (sttts) [inline hack/local-up-cluster.sh]: How about also printing to the user how to use this aggregator config?

File: hack/local-up-cluster.sh (hunk 6 above)
Comment (deads2k) [inline hack/local-up-cluster.sh]: need a namespace to create the pod in.  In order to be publicly accessible, `kube-public` seems like a reasonable spot.

File: hack/local-up-cluster.sh (hunk 3 above)
Comment (sttts) [inline hack/local-up-cluster.sh]: I prefer `/var/run/kubernetes/static-pods` over some fixed /tmp dir.

File: cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml [hunk 8]
```diff
@@ -0,0 +1,81 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: kube-aggregator
+  namespace: kube-public
```
Comment (sttts) [inline cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml]: Would have expected kube-system. Other addons go under kube-systems. Any reason that the aggregator doesn't?

File: cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+spec:
+  hostNetwork: true
+  containers:
+  - name: kube-aggregator
+    image: kube-aggregator
```
Comment (sttts) [inline cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml]: better version this.

File: cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    - "--proxy-client-key-file=/var/run/auth-proxy-client/client-auth-proxy.key"
+    - "--tls-cert-file=/var/run/serving-cert/serving-kube-aggregator.crt"
+    - "--tls-private-key-file=/var/run/serving-cert/serving-kube-aggregator.key"
+    - "--tls-ca-file=/var/run/serving-ca/server-ca.crt"
+    - "--client-ca-file=/var/run/client-ca/client-ca.crt"
+    - "--requestheader-username-headers=X-Remote-User"
+    - "--requestheader-group-headers=X-Remote-Group"
+    - "--requestheader-extra-headers-prefix=X-Remote-Extra-"
+    - "--requestheader-client-ca-file=/var/run/request-header-ca/request-header-ca.crt"
+    - "--etcd-servers=http://127.0.0.1:2379"
```
Comment (sttts) [inline cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml]: this is the host etcd with `hostNetwork: true`. Intentional?

Comment (deads2k) [inline cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml]: yes

File: cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml (hunk 9 above)
Comment (deads2k) [inline cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml]: > better version this.

This isn't the final image.

File: cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml (hunk 8 above)
Comment (deads2k) [inline cmd/kube-aggregator/artifacts/hostpath-pods/insecure-etcd-pod.yaml]: > Would have expected kube-system. Other addons go under kube-systems. Any reason that the aggregator doesn't?

That's a change that will need a wider discussion.

File: hack/local-up-cluster.sh (hunk 7 above)
Comment (deads2k) [inline hack/local-up-cluster.sh]: > How about also printing to the user how to use this aggregator config?

done.
